modules/core/src/smithy4s/Blob.scala

```diff
@@ -18,101 +18,238 @@ package smithy4s
 
-  def apply(bytes: Array[Byte]): Blob = new ByteArrayBlob(bytes)
-  def apply(buffer: ByteBuffer): Blob = new ByteBufferBlob(buffer)
-  def apply(string: String): Blob = new ByteArrayBlob(
-    string.getBytes(StandardCharsets.UTF_8)
-  )
+  def apply(bytes: Array[Byte]): Blob = new ArraySliceBlob(bytes, 0, bytes.length)
+  def apply(string: String): Blob = apply(string.getBytes(StandardCharsets.UTF_8))
+  def apply(buffer: ByteBuffer): Blob = view(buffer.duplicate())
 
-  trait Encoder[A] {
-    def encode(a: A): Blob
-  }
+  def slice(bytes: Array[Byte], offset: Int, size: Int): Blob = new ArraySliceBlob(bytes, offset, size)
+  def view(buffer: ByteBuffer): Blob = new ByteBufferBlob(buffer)
+  def queue(blobs: Queue[Blob], size: Int) = new QueueBlob(blobs, size)
 
-  trait Decoder[E, A] {
-    def decode(blob: Blob): Either[E, A]
-  }
+  final class ByteBufferBlob private[smithy4s] (val buf: ByteBuffer) extends Blob {
+    def apply(i: Int) = buf.get(i.toInt)
 
-  private final class ByteArrayBlob(private val bytes: Array[Byte])
-      extends Blob {
-    override def toArray: Array[Byte] = bytes
-    override def isEmpty: Boolean = bytes.isEmpty
-    override def size: Int = bytes.length
-    override def buffer: ByteBuffer =
-      ByteBuffer.wrap(java.util.Arrays.copyOf(bytes, bytes.length))
+    override def toArraySliceBlob: ArraySliceBlob = if (buf.hasArray()) {
+      new ArraySliceBlob(buf.array, buf.arrayOffset, size)
+    } else super.toArraySliceBlob
 
-    override def toString = {
-      s"ByteArrayBlob[${Base64.getEncoder().encodeToString(bytes)}]"
+    override def copyToArray(xs: Array[Byte], start: Int, offset: Int, size: Int): Unit = {
+      val n = buf.duplicate()
+      n.position(offset.toInt)
+      n.get(xs, start, size)
+      ()
     }
 
-    override def sameBytesAs(other: Blob): Boolean = other match {
-      case otherBlob: ByteArrayBlob =>
-        java.util.Arrays.equals(bytes, otherBlob.bytes)
-      case otherBlob: ByteBufferBlob =>
-        ByteBuffer.wrap(bytes).compareTo(otherBlob.buffer) == 0
+    override def toArray: Array[Byte] = {
+      val arr = Array.ofDim[Byte](buf.remaining())
+      copyToArray(arr, 0, 0, size)
+      arr
     }
 
-    override def equals(other: Any): Boolean = {
-      other.isInstanceOf[ByteArrayBlob] &&
-      java.util.Arrays.equals(bytes, other.asInstanceOf[ByteArrayBlob].bytes)
-    }
+    override def asByteBuffer(offset: Int, size: Int): ByteBuffer =
+      asByteBufferUnsafe(offset, size).asReadOnlyBuffer()
 
-    override def hashCode(): Int = {
-      var hashCode = 0
-      var i = 0
-      while (i < bytes.length) {
-        hashCode += bytes(i).hashCode()
-        i += 1
+    override def asByteBufferUnsafe(offset: Int, size: Int): ByteBuffer = {
+      val b = buf
+      if (offset == 0 && b.position() == 0 && size == b.remaining()) b
+      else {
+        b.position(offset.toInt)
+        b.limit(offset.toInt + size)
+        b.slice()
       }
```

Comment:
@daddykotex see here.

Reply:
ahhh I was looking at the default implementation, and I did not realize there was an override for it